### PR TITLE
add user setting to allow users to stay subscribed until manual unsubscribe

### PR DESCRIPTION
**Pull request by shannon-heh**
**Goal**: Based on user feedback, some users request that we allow them to remain subscribed to a section (until they manually unsubscribe), instead of auto-unsubscribing them after first notification. This change implements the "stay subscribed" feature and allows users to choose which setting they prefer.

**Terminology Clarification:**

- auto-resubscribed = auto-resub = stay subscribed
- auto-unsubscribe is opposite of auto-resubscribe

**Major changes:**

- Add `auto_resub` flag for each user in DB: `True` if they want to stay subscribed, `False` (default) if they want to auto-unsub. Backwards compatible: flag does not appear for current users, but flag is added if they toggle switch once.
- Add switch to user settings modal: toggle switch ON to stay subscribed, OFF to auto-unsub after first notification. Includes info tooltip to clarify purpose. Alert appears to confirm setting change is successful.
- Adjust body of email and text message based on whether user is auto-resubbed

**Demos:**

_Toggling the switch changes `auto_resub` flag for user in DB:_

https://user-images.githubusercontent.com/63625700/138027079-5a458359-0b97-450b-bd87-5a0cc6c20009.mov

_When `auto_resub` is `True`, sending notif does NOT unsub user:_

https://user-images.githubusercontent.com/63625700/138028791-242bbfb5-e517-4c11-8db5-b4642a7506d8.mov

_When `auto_resub` is `False`, sending notif does unsub user:_

https://user-images.githubusercontent.com/63625700/138029035-b5d38e0a-9dea-4e1a-a2d0-28d28ee2c30a.mov

_Email and text message bodies are different based on whether user is auto-resubbed_

<table>
<tr>
<td>
Screen Shot 2021-10-19 at 10 05 08 PM
</td>
<td>
Screen Shot 2021-10-19 at 10 05 22 PM
</td>
</tr>
</table>

Screen Shot 2021-10-19 at 9 43 55 PM

**Tests:**
- Unit testing done on `database.py` methods (and checking changes in DB).
- Manual testing done with local app + staging DB: manually filled a section (using admin button), turned auto-resub ON, then run `send_notifs.py` (trigger notifs script) more than once to see that user is not auto-unsubbed.

**Potential Issues:**
- Since we don't set a max # emails sent for a given section, users may find receiving the same email over and over again to be an annoyance (especially if a spot opens at a time when a user isn't accessing their email/texts, e.g. when they're sleeping).
- UI for switch row is inconsistent with UI for other form fields :(

**Future Work:**
- Allow users to toggle how many emails we send before auto-unsubscribing (e.g. 5, 10 emails).

### CODE REVIEW

File: src/notify.py [hunk 1]
```diff
@@ -59,33 +59,37 @@ def get_name(self):
     # sends a formatted email
 
     def send_emails_html(self):
+        next_step_unsubbed = f"""<p>You've been <b>automatically unsubscribed</b> from this section. If you didn't get the spot, you may re-subscribe here: <a href="https://snatch.tigerapps.org/course?query=&courseid={self._courseid}&skip">TigerSnatch | {self._deptnum}</a>.</p>"""
```
Comment (nicholaspad) [inline src/notify.py:62]: We should let users know that they can change their auto-unsubscribe preferences here. I would append the sentence:

> To turn off auto-unsubscription, click Edit Account Settings on the TigerSnatch Dashboard.

File: src/notify.py [hunk 2]
```diff
@@ -59,33 +59,37 @@ def get_name(self):
...
+
+        next_step_resubbed = f"""<p>To stop receiving notifications for this section, unsubscribe here: <a href="https://snatch.tigerapps.org/course?query=&courseid={self._courseid}&skip">TigerSnatch | {self._deptnum}</a>.</p>"""
```
Comment (nicholaspad) [inline src/notify.py:64]: Ditto:

> Turn on auto-unsubscription using Edit Account Settings on the TigerSnatch Dashboard.

File: src/notify.py [hunk 3]
```diff
@@ -59,33 +59,37 @@ def get_name(self):
...
         msg = f"""\
         <html>
         <head></head>
         <body style='font-size:1.3em'>
             <p>Greetings $$netid$$,</p>
             <p>Your subscribed section <b>{self._sectionname}</b> in <b>{self._coursename}</b> has one or more spots open! Note that some courses reserve spots or are closed, so enrollment may not be possible.</p>
             <p>Head over to <a href="https://phubprod.princeton.edu/psp/phubprod/?cmd=start">TigerHub</a> to Snatch your spot!</p>
-            <p>You've been <b>automatically unsubscribed</b> from this section. If you didn't get the spot, you may re-subscribe here: <a href="https://snatch.tigerapps.org/course?query=&courseid={self._courseid}&skip">TigerSnatch | {self._deptnum}</a>.</p>
+            $$next_step$$
             <p>To receive text message notifications about open spots, <b>set your phone number</b> on the <a href="https://snatch.tigerapps.org/dashboard">TigerSnatch Dashboard</a> if you haven't already!</p>
```
Comment (nicholaspad) [inline src/notify.py:74]: We should let users know about auto-unsubscription (and text message notifications) in the new Edit Account Settings panel. I suggest replacing the sentence with this:

> To enable/disable auto-unsubscription or set your phone number for SMS notifications, use <b>Edit Account Settings</b> on the <a href="https://snatch.tigerapps.org/dashboard">TigerSnatch Dashboard</a>!

File: src/notify.py [hunk 4]
```diff
@@ -59,33 +59,37 @@ def get_name(self):
...
-        }
-
+        
         try:
+            data = {
+                "personalizations": [
+                    {
+                        "to": [{"email": self._emails[i]}],
+                        "subject": f"TigerSnatch: a spot opened in {self._deptnum} {self._sectionname}",
+                        "substitutions": {"$$netid$$": self._netids[i], "$$next_step$$": (next_step_resubbed if self.db.get_user_auto_resub(self._netids[i]) else next_step_unsubbed)},
```
Comment (nicholaspad) [inline src/notify.py:85]: I think it'd be more readable to use a separate variable for the call to `get_user_auto_resub` like you do in line 106 below:

```
is_auto_resub = self.db.get_user_auto_resub(self._netids[i])
```

File: src/notify.py [hunk 5]
```diff
@@ -95,14 +99,17 @@ def send_emails_html(self):
     # sends an SMS
 
     def send_sms(self):
-        msg = f"{self._sectionname} in {self._deptnum} has open spots! You've been unsubscribed from this section. Resubscribe here: https://snatch.tigerapps.org/course?query=&courseid={self._courseid}&skip"
+        msg_unsubbed = f"{self._sectionname} in {self._deptnum} has open spots! You've been unsubscribed from this section. Resubscribe here: https://snatch.tigerapps.org/course?query=&courseid={self._courseid}&skip"
+        msg_resubbed = f"{self._sectionname} in {self._deptnum} has open spots! To stop receiving notifications for this section, unsubscribe here: https://snatch.tigerapps.org/course?query=&courseid={self._courseid}&skip"
```
Comment (nicholaspad) [inline src/notify.py:103]: Text messages via SMS have a 160 character limit. `msg_resubbed` can actually go up to 163 characters, for example:

> P01A in CHI1001 has open spots! To stop receiving notifications for this section, unsubscribe here: https://snatch.tigerapps.org/course?query=&courseid=123456&skip

In particular, this text has a letter in the section name and a 4-digit course number. We need to cut off >=4 characters from this string to be safe. To do this, I suggest removing `query=&` from the URL (example: https://snatch.tigerapps.org/course?courseid=002051&skip). This should also be done for the URL in `msg_unsubbed`.

Comment (nicholaspad) [inline src/notify.py:103]: Text messages via SMS have a 160 character limit. `msg_resubbed` can actually go up to 163 characters, for example:

> P01A in CHI1001 has open spots! To stop receiving notifications for this section, unsubscribe here: https://snatch.tigerapps.org/course?query=&courseid=123456&skip

In particular, this edge case has a letter in the section name and a 4-digit course number. We need to cut off >=4 characters from this string to be safe. To do this, I suggest removing `query=&` from the URL (example: https://snatch.tigerapps.org/course?courseid=002051&skip). This should also be done for the URL in `msg_unsubbed`.

File: static/app.js [hunk 6]
```diff
@@ -614,6 +616,26 @@ let switchListener = function () {
...
 
+let autoResubSwitchListener = function() {
+  $("#auto-resub-switch").change(function(e) {
+    e.preventDefault();
+    disableSwitchFunctions();
+    const checkedProp = $("#auto-resub-switch").prop("checked");
+    $.post(`/update_auto_resub/${checkedProp}`, function(res) {
+      if (res["isSuccess"]) {
+        if (checkedProp) {
+          alert("Notification settings successfully changed: You will keep receiving open spot notifications for a section until you manually unsubscribe.")
```
Comment (nicholaspad) [inline static/app.js:627]: nit: change "open spot" to "open-spot"

File: views/utils/user_settings.html [hunk 7]
```diff
@@ -81,6 +81,47 @@
...
+                                viewBox="0 0 16 20"
+                                data-bs-toggle="tooltip"
+                                data-bs-placement="top"
+                                title="Turn on the switch to keep receiving open spot notifications for a section until you manually unsubscribe. Turn off the switch to be auto-unsubscribed upon first notification.">
+                                <path
+                                    d="M0 2a2 2 0 0 1 2-2h12a2 2 0 0 1 2 2v12a2 2 0 0 1-2 2H2a2 2 0 0 1-2-2V2zm8.93 4.588-2.29.287-.082.38.45.083c.294.07.352.176.288.469l-.738 3.468c-.194.897.105 1.319.808 1.319.545 0 1.178-.252 1.465-.598l.088-.416c-.2.176-.492.246-.686.246-.275 0-.375-.193-.304-.533L8.93 6.588zM8 5.5a1 1 0 1 0 0-2 1 1 0 0 0 0 2z" />
+                                </svg>
+                            </label>
+                            <div>
+                            {% if auto_resub == True %}
```
Comment (nicholaspad) [inline views/utils/user_settings.html:107]: nit: `if auto_resub`

File: views/utils/user_settings.html [hunk 8]
```diff
@@ -81,6 +81,47 @@
...
+                                <div class="form-check form-switch">
+                                    <input class="form-check-input"
+                                        type="checkbox"
+                                        id="auto-resub-switch"
+                                        checked />
+                                </div>
+                            {% elif auto_resub == False %}
```
Comment (nicholaspad) [inline views/utils/user_settings.html:114]: nit: `else`

(not sure if this change and the above comment's change will work)

File: src/database.py [hunk 9]
```diff
@@ -847,6 +848,27 @@ def get_current_sections(self, netid):
...
         return res
 
+    # sets auto resubscribe flag for user
+
+    def update_user_auto_resub(self, netid, auto_resub):
+        try:
+            self._db.users.update_one({"netid": netid}, {"$set": {"auto_resub": auto_resub}})
+            return True
+        except:
+            print(f"attempt to update auto_resub for {netid} failed")
```
Comment (nicholaspad) [inline src/database.py:858]: add `file=stderr` to the print statement:

```
print(f"attempt to update auto_resub for {netid} failed", file=stderr)
```

**Review by nicholaspad — CHANGES_REQUESTED** [review]
Amazing work on this :) everyone will be happy that this is finally a feature 🥳 I tested the feature on the staging app and found no issues. I left a few suggestions (mostly about wording and grammar) as inline comments!

File: views/utils/user_settings.html (hunk 7 above)
Comment (shannon-heh) [reply views/utils/user_settings.html:107]: meant to change this, cuz i did `== True` and `== False` to be extra safe

**Pull request merged**